### PR TITLE
[BI-982] - Ontology BrAPI Mapping for Trait

### PR DESCRIPTION
**Pull request by dmeidlin**
The traitName for BrAPI traits is set to entity + attribute.
The traitName for BI traits is renamed to observationVariableName and is set to the value under the "Ontology term name" column for trait import.
Method names are autoformed from description + class

With these changes in place trait creation and editing will no longer work from the UI. Traits can only be added by batch importing until the UI for traits is updated.

Trait File Import:
- added "Trait description", "Trait entity", and "Trait attribute" columns
- renamed "Trait name" to "Ontology term name"
- added validators for the additional columns

Trait model:
- traitDescription field added

BI db:
- trait_name column in the trait table renamed to observation_variable_name(no data migration necessary)
- src code and tests updated to use new field name for traits.

### CODE REVIEW

**Review by timparsons — APPROVED** [review]
Looks good! Approving for the current code, but depending on the decision of the comment left on https://github.com/Breeding-Insight/bi-web/pull/106

**Comment by nickpalladino** [comment]
> With these changes in place trait creation and editing will no longer work from the UI. Traits can only be added by batch importing until the UI for traits is updated.

Should we hold off on merging this card after it is approved until that card that addresses the UI side is done so we don't have broken traits on develop?

File: src/main/java/org/breedinginsight/services/parsers/trait/TraitFileParser.java
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/parsers/trait/TraitFileParser.java:128]: Is entity really the same as observation level or is it just being used for the autocomplete features of the program observation levels?

**Comment by timparsons** [comment]
> > With these changes in place trait creation and editing will no longer work from the UI. Traits can only be added by batch importing until the UI for traits is updated.
> 
> Should we hold off on merging this card after it is approved until that card that addresses the UI side is done so we don't have broken traits on develop?

Probably not a bad idea

File: src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java:78]: Relating to previous comment, if the ObservationVariable was changed to have a trait object property then these would be trait.entity, trait.attribute, and trait.traitDescription to be consistent with how the other messages are formatted with the first part being the brapi object the properties are under. But that could be something done in another card.

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/model/Trait.java:59]: If this class is really ObservationVariable, it may make sense to move the brapi trait object properties to a Trait object property like is being done for method and scale. Could be done in another card.

File: src/main/java/org/breedinginsight/services/parsers/trait/TraitFileParser.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/services/parsers/trait/TraitFileParser.java:128]: just for the autocomplete features.
The card indirectly states that the two are not the same by saying that obs level should not be associated with traits.

File: src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java:78]: agreed. There needs to be a major refactoring done(separate card) on the backend to go from using Trait as the highest level object (as it has been) to using observation variable as the highest level object(as you describe) to be consistent with the brapi model.
The scope of this work is functinality: making the changes necessary to get the backend to support the MVP changes made on the frontend. In the spirit of breaking the work up into manageable pieces done in a single sprint, I think the remaining necessary refactoring should be a separate card.

File: src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java:151]: I believe it's correct as is.

In the Ontology MVP changes entity + attribute is mapped to the name of the BrAPI trait object, and the ontology term name (or just Name) is mapped to the BrAPI observation variable object, so 
observationVariable.trait.traitName = entity + attribute
observationVariable.observationVariableName = Name (which is stored in the Trait model as observationVariableName).

File: src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java]: Good catch! Yes, OV name wasn't intentional and should be description: fixed and pushed up.

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/model/Trait.java:82]: agreed. done.

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/model/Trait.java:59]: agreed. part of what I imagined would be included in the refactoring mentioned above.

File: src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/brapi/v1/services/BrapiObservationVariableService.java:151]: I looked at it again, you're right. line 151 is setting observationVariable.trait.traitName. will make the change.
Also will change line 146 should to `.description(trait.getTraitDescription).`

File: src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/services/validators/TraitValidatorError.java:78]: sounds good, could you create the card for that or at least a placeholder for it

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/model/Trait.java:59]: sounds good

File: src/main/java/org/breedinginsight/services/parsers/trait/TraitFileColumns.java
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/services/parsers/trait/TraitFileColumns.java:25]: Talked with @shawnyarnes and this will be addressed in BI-1059

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/model/Trait.java:111]: There is a separate card for this: [BI-996](https://breedinginsight.atlassian.net/browse/BI-996).

File: src/main/java/org/breedinginsight/model/Trait.java
Comment (dmeidlin) [reply src/main/java/org/breedinginsight/model/Trait.java:112]: Thanks, agreed.

**Review by nickpalladino — APPROVED** [review]

**Pull request merged**
